src/wallet.lisp

```diff
@@ -18,39 +16,133 @@
+(defvar *default-wallet-name* "My Wallet")
+
+(defun create-wallet (&key
+                        (name *default-wallet-name*)
+                        (force nil))
+  "Create an on disk wallet with NAME
+
+If the wallet already exists, it will not be overwritten unless FORCE
+is non-nil.
+
+The default name for a wallet is *DEFAULT-WALLET-NAME*."
+  (let ((path (emotiq-wallet-path :name name)))
+    (when (and (probe-file path)
+               (not force))
+      (format *standard-output* "Not overwriting existing wallet at '~a'." path)
+      (return-from create-wallet nil))
+    (let ((wallet (make-wallet)))
+      (values
+       name
+       (wallet-serialize wallet :path path)))))
+
+(defun get-wallet-named (name)
+  (wallet-deserialize :path (emotiq-wallet-path :name name)))
+
+(defun rename-wallet (from-name to-name)
+  "Rename wallet FROM-NAME into TO-NAME
+
+Returns nil if unsuccessful."
+  (let ((from (emotiq-wallet-path :name from-name))
+        (to (emotiq-wallet-path :name to-name)))
+    (unless (probe-file from)
+      (format *standard-output* "Not renaming.~&No such wallet named '~a' already exists as '~a'."
+              from-name
+              from)
+      (return-from rename-wallet nil))
+    (when (probe-file to)
+      (format *standard-output* "Not renaming.~&A wallet named '~a' already exists as '~a'."
+              to-name
+              to)
+      (return-from rename-wallet nil))
+    (ensure-directories-exist to)
+    (prog1
+        (rename-file from to)
+      (uiop:delete-empty-directory
+       (make-pathname :name nil
+                      :type nil
+                      :defaults from)))))
+
+(defun enumerate-wallets ()
+  "Return a list of wallet names persisted on the local node"
+  (let ((directories
+         (directory 
+          (merge-pathnames "../*/"
+                           (make-pathname :name nil :type nil
+                                          :defaults (emotiq-wallet-path))))))
+    (loop
+       :for directory :in directories
+       :collecting (pathname-name-to-wallet-name
+                    (first (last (pathname-directory directory)))))))
+
+;;; We want to allow "human readable" wallet names that have a
+;;; one-to-one mapping to valid directory names.  Using URI encoding
+;;; ensures that we can have whitespace and slash characters in the
+;;; filename.
+(defun wallet-name-to-pathname-name (name)
+  (quri:url-encode name))
+
+(defun pathname-name-to-wallet-name (name)
+  (quri:url-decode name))
+
+(defun emotiq-wallet-path (&key (name *default-wallet-name*))
+  "Return pathname of wallet with NAME"
+  (merge-pathnames
+   (format nil "wallet/~a/~a"
+           (wallet-name-to-pathname-name name)
+           "emotiq.wallet")
+   (emotiq/user/root)))
    
 (defun wallet-serialize (wallet &key (path (emotiq-wallet-path)))
+  "Serialize WALLET object to PATH"
   (ensure-directories-exist path)
   (with-open-file
       (o path :direction :output :element-type '(unsigned-byte 8))
```

Comment:
- There are references to unexported symbols (using double colons ::), e.g., in hexify method. Please eliminate (or explain?)
- What's the magic 31 value in key-phrase? Can the be a comment at least, or a variable or constant naming it?
- there's no :if-exists handler in the with-open-file in wallet-serialize. what happens if the file exists? some lisps and/or os's could cause an error. if the intent is to supersede, should make that explicit.  Or can one backup be kept?  EIther way a staged supersede is probably best: a backup should be kept at least until the writing is done, then it could be gotten rid of.  If the writing is interrupted in the middle, you should not be left with nothing.